### DISCUSSION TITLE
To state or not to state?

### DISCUSSION THREAD

**Discussion by giorgosera** [serverlessworkflow/specification#660] [Q&A] [answered]
Hi there, I'm super new to Serverless DSL and I have the following question. 

In the example ['Event based greeting'](https://github.com/serverlessworkflow/specification/tree/main/examples#Event-Based-Greeting-Example) there is a single state that is triggered by an event which then executes an action `greetingFunction `. 

I was wondering if the following alternative would be acceptable and if not why?

- Instead of one state I can have two states. 
- The first is an Event state which references the greeting event and then immediately transitions to the next state without executing the action.
- Then in the second state (operation state) I execute the action `greetingFunction`.

I guess my question is when should I group my actions under one state vs spreading them to different states? 
e.g. Why would I have an operation state with two actions AFTER an event state when I can put those actions under the event state and get rid of the operation state? I am failing to understand when to choose the one over the other.

It feels like I am misunderstanding something very fundamental so I hope someone can help me!

Thank you

**Comment by ricardozanini** [comment, marked as answer]
Hi there! Sorry for the long winter I hope you can still see value on my answer.

> I guess my question is when should I group my actions under one state vs spreading them to different states?
e.g. Why would I have an operation state with two actions AFTER an event state when I can put those actions under the event state and get rid of the operation state? I am failing to understand when to choose the one over the other.

There's no correct answer. You can group actions in a state when it makes sense to you. For example, when the actions complement each other. 

In the new DSL, each task is responsible for a single operation:

```yaml
do:
  - task1
  - task2
  - task3
```

In your example, you would have a `listen` task before calling the `greetingFunction`.

Check out our example directories to get an idea: https://github.com/serverlessworkflow/specification/tree/main/examples
